internal/oci/notation/notation.go

```diff
@@ -0,0 +1,319 @@
+	return func(o *options) {
+		o.Keychain = key
+	}
+}
+
+// WithLogger is a function that returns an Options function to set the logger for the options.
+// The logger is used for logging purposes within the options.
+func WithLogger(logger logr.Logger) Options {
+	return func(o *options) {
+		o.Logger = logger
+	}
+}
+
+// NotationVerifier is a struct which is responsible for executing verification logic
+type NotationVerifier struct {
+	auth     authn.Authenticator
+	keychain authn.Keychain
+	verifier *notation.Verifier
+	opts     []remote.Option
+	insecure bool
+	logger   logr.Logger
+}
+
+type trustStore struct {
+	cert []byte
+}
+
+// GetCertificates implements truststore.X509TrustStore.
+func (s trustStore) GetCertificates(ctx context.Context, storeType truststore.Type, namedStore string) ([]*x509.Certificate, error) {
+	raw := s.cert
+	block, _ := pem.Decode(raw)
+	if block != nil {
+		raw = block.Bytes
+	}
+
+	certs, err := x509.ParseCertificates(raw)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse certificate '%s': %s", namedStore, err)
+	}
+
+	return certs, nil
+}
+
+// NewNotationVerifier initializes a new Verifier
+func NewNotationVerifier(opts ...Options) (*NotationVerifier, error) {
+	o := options{}
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	store := &trustStore{
+		cert: o.RootCertificate,
+	}
+
+	verifier, err := verifier.New(o.TrustPolicy, store, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &NotationVerifier{
+		auth:     o.Auth,
+		keychain: o.Keychain,
+		verifier: &verifier,
+		opts:     o.ROpt,
+		insecure: o.Insecure,
+		logger:   o.Logger,
+	}, nil
+}
+
+// Verify verifies the authenticity of the given ref OCI image.
+// It returns a boolean indicating if the verification was successful.
+// It returns an error if the verification fails, nil otherwise.
+func (v *NotationVerifier) Verify(ctx context.Context, ref name.Reference) (oci.VerificationResult, error) {
+	url := ref.Name()
+
+	remoteRepo, err := v.remoteRepo(url)
+	if err != nil {
+		return oci.VerificationResultFailed, err
+	}
+
+	repo := registry.NewRepository(remoteRepo)
+
+	repoUrl, err := v.repoUrlWithDigest(url, ref)
+	if err != nil {
+		return oci.VerificationResultFailed, err
+	}
+
+	verifyOptions := notation.VerifyOptions{
+		ArtifactReference:    repoUrl,
+		MaxSignatureAttempts: 3,
+	}
+
+	_, outcomes, err := notation.Verify(ctx, *v.verifier, repo, verifyOptions)
+	if err != nil {
+		return oci.VerificationResultFailed, err
+	}
+
+	if len(outcomes) == 0 {
+		return oci.VerificationResultFailed, fmt.Errorf("signature verification failed for all the signatures associated with %s", url)
+	}
+
+	outcome := outcomes[0]
+
+	if outcome.VerificationLevel == trustpolicy.LevelSkip {
+		return oci.VerificationResultIgnored, nil
+	}
+
+	for _, i := range outcome.VerificationResults {
+		if i.Error != nil {
+			if i.Type == trustpolicy.TypeAuthenticity {
+				return oci.VerificationResultIgnored, i.Error
```

Comment:
I think the above code is correct.

The way I understood it, based on https://github.com/notaryproject/specifications/blob/main/specs/trust-store-trust-policy.md#signature-verification-details:
1. for `strict` and `permissive` levels, types `integrity` and `authenticity` are enforced. This means that an error will be returned if the fail by `notation.Verify(ctx, *v.verifier, repo, verifyOptions)` and we can return the `error` directly without looking a the `outcome`
2. for `skip` everything is skipped, so we can just return `ignored`
3. At this point if we haven't returned, there is only one level to check `audit`. For this level only `integrity` is enforced (which means it return an error directly for that type). Every other type is logged. But in our case we want to return a `successful` verification if `authenticity` is successful and `ignored` otherwise.

Reply:
Apologies for the slow response. I should have documented what was happening as it did take me some time to figure out the best approach for handling the various scenarios mentioned previously.

I will split this out to its own function and add additional unit testing around it. I will also update the for loop block so that it doesn't just immediately return if error occurred, and type is `authenticity`. 

Something like:

```go
ignore := false

for _, i := range outcome.VerificationResults {
	if i.Error != nil {
		if i.Type == trustpolicy.TypeAuthenticity {
			ignore = true
		}

		v.logger.Info(fmt.Sprintf("verification check for type %s failed for %s with message %s", i.Type, url, i.Error.Error()))
	}
}

if ignore {
	return oci.VerificationResultIgnored, nil
}

return oci.VerificationResultSuccess, nil
```

That way the other types can be logged and still return a `VerificationResultIgnored` so that the artifact is not marked as verified but still allowed to be deployed.

In terms of how it works, @souleb you are correct. Any type marked as `enforced` will immediately fail in `notation.Verify(ctx, *v.verifier, repo, verifyOptions)`, `outcome` will be `nil` and there will be an `err` which will result in `VerificationResultFailed` and return with the `err`.

If the `VerificationLevel` is set to `skip` then `VerificationResults` in `outcomes` will be `nil` since nothing was checked. `skip` cannot be customise either via an `override` (doing so will result in a `trust policy statement "test-statement-name" has invalid signatureVerification: signature verification level "skip" can't be used to customize signature verification`) and would have failed long before we've gotten to this stage. So, all we need to do is check `VerificationLevel` to see if it is set to `skip` and if so, return with `VerificationResultIgnored`.

Next, the for loop block goes through the `outcome` array and checks for `Error` at each of the different validation steps (`integrity`, `authenticity`, `authenticTimestamp`, etc.) The `Error` attribute will only contain an `error` if that step failed and the action is `log`. If it is marked as `skip` (because you customised the verification), that step will not be included in the result and if it passed that step, `Error` will be `nil`. If it is marked as `enforced`, it would have already failed before getting to this point.

Reply:
I should add that it appears `outcomes` should only be of size 1 if verification is successful. 

As shown here (https://github.com/notaryproject/notation-go/blob/7fa8404e79d95dfb8ff0a89754dcdc57cea6945b/notation.go#L376C4-L395C30) if the artifact contains multiple signatures (either because the artifact was signed by different certs, the certs were revoked, expired, etc.) then notation checks every signature associated with the artifact. Each signature is verified and a new `outcome` is created (https://github.com/notaryproject/notation-go/blob/7fa8404e79d95dfb8ff0a89754dcdc57cea6945b/notation.go#L376). 

If there's an error then `outcome.Error` is set but then it is used to populate `verificationFailedErrorArray` (https://github.com/notaryproject/notation-go/blob/7fa8404e79d95dfb8ff0a89754dcdc57cea6945b/notation.go#L384) before continuing to the next signature. It is only when there's no error `outcome` is used (with no outcome.Error), is added to `verificationOutcomes` (https://github.com/notaryproject/notation-go/blob/7fa8404e79d95dfb8ff0a89754dcdc57cea6945b/notation.go#L391) and the loop exits.

Either `verificationFailedErrorArray` (with many errors) or `verificationOutcomes` (with just the single outcome) is then used (unless `MaxSignatureAttempts` is met). This is why we can just take the first item in `outcome := outcomes[0]`

Reply:
Applied the changes in [c3a89f4](https://github.com/fluxcd/source-controller/pull/1075/commits/c3a89f4025768101f0d9ae813bb1916a53959cb3). Added additional comments in [0dedf6f](https://github.com/fluxcd/source-controller/pull/1075/commits/0dedf6f098a9897a3bc2e54306d77af6bf15b449) as well.

Reply:
Resolved in 3743903

Reply:
Resolved in 7249666, sorry i should have reviewed it thoroughly